REGRESSION (251792@main?): [ macOS ][ iOS ] TestWebKitAPI.WKBackForwardList.BackForwardNavigationDoesNotSkipItemsWithUserGesture is a flaky failure and crash

cdumez opened the pull request:
#### 2a7a51012dfcd5468c4e5ad6a41654ade58fe39e
<pre>
REGRESSION (251792@main?): [ macOS ][ iOS ] TestWebKitAPI.WKBackForwardList.BackForwardNavigationDoesNotSkipItemsWithUserGesture is a flaky failure and crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=241991">https://bugs.webkit.org/show_bug.cgi?id=241991</a>
&lt;rdar://95886849&gt;

Reviewed by Alex Christensen.

The API test relies on `[webView _test_waitForDidFinishNavigationOrSameDocumentNavigation]` to proceed
to the next steps instead of the usual `[webView _test_waitForDidFinishNavigation]` because
didFinishNavigation does not get called for same document navigations (e.g. back navigation after
history.pushState()).

However, what I didn&apos;t realize is that when doing a back navigation after a pushState(),
didSameDocumentNavigation would get called twice, once with _WKSameDocumentNavigationTypeAnchorNavigation
and then again with _WKSameDocumentNavigationTypeSessionStatePop. These duplicate calls would cause the
test to proceed further too eagerly.

To address the issue, these API tests now rely on their own custom navigation delegate which only marks
the navigation as complete if didSameDocumentNavigation gets called with either
_WKSameDocumentNavigationTypeSessionStatePush or _WKSameDocumentNavigationTypeSessionStatePop.

* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardList.mm:
(-[WKBackForwardNavigationDelegate init]):
(-[WKBackForwardNavigationDelegate webView:didFinishNavigation:]):
(-[WKBackForwardNavigationDelegate _webView:navigation:didSameDocumentNavigation:]):
(TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:navigation:didSameDocumentNavigation:]): Deleted.
(-[TestNavigationDelegate waitForDidFinishNavigationOrSameDocumentNavigation]): Deleted.
(-[WKWebView _test_waitForDidFinishNavigationOrSameDocumentNavigation]): Deleted.

Canonical link: <a href="https://commits.webkit.org/251894@main">https://commits.webkit.org/251894@main</a>
</pre>

achristensen07 commented on Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
```diff
@@ -100,7 +100,7 @@ - (void)webView:(WKWebView *)webView didReceiveAuthenticationChallenge:(NSURLAut
 - (void)_webView:(WKWebView *)webView navigation:(WKNavigation *)navigation didSameDocumentNavigation:(_WKSameDocumentNavigationType)navigationType
 {
-    if (_didSameDocumentNavigation)
+    if (_didSameDocumentNavigation && navigationType != _WKSameDocumentNavigationTypeAnchorNavigation)
```
TestNavigationDelegate's job is to be thin and prevent each test from needing its own class declaration.  I think this logic would be better placed in the test.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardList.mm:
```diff
@@ -311,6 +312,32 @@ - (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigati
     EXPECT_STREQ([[list.currentItem URL] absoluteString].UTF8String, url3.absoluteString.UTF8String);
 }
 
+static bool navigated;
```
It would be better to not introduce a new static global boolean.  Can we make this an ivar on the delegate and introduce a waitForSomething method on the delegate instead?

cdumez replied:
Ok, done.

cdumez committed:
REGRESSION (251792@main?): [ macOS ][ iOS ] TestWebKitAPI.WKBackForwardList.BackForwardNavigationDoesNotSkipItemsWithUserGesture is a flaky failure and crash
https://bugs.webkit.org/show_bug.cgi?id=241991
<rdar://95886849>

Reviewed by Alex Christensen.

The API test relies on `[webView _test_waitForDidFinishNavigationOrSameDocumentNavigation]` to proceed
to the next steps instead of the usual `[webView _test_waitForDidFinishNavigation]` because
didFinishNavigation does not get called for same document navigations (e.g. back navigation after
history.pushState()).

However, what I didn't realize is that when doing a back navigation after a pushState(),
didSameDocumentNavigation would get called twice, once with _WKSameDocumentNavigationTypeAnchorNavigation
and then again with _WKSameDocumentNavigationTypeSessionStatePop. These duplicate calls would cause the
test to proceed further too eagerly.

To address the issue, these API tests now rely on their own custom navigation delegate which only marks
the navigation as complete if didSameDocumentNavigation gets called with either
_WKSameDocumentNavigationTypeSessionStatePush or _WKSameDocumentNavigationTypeSessionStatePop.

* Tools/TestWebKitAPI/Tests/WebKit/WKBackForwardList.mm:
(-[WKBackForwardNavigationDelegate init]):
(-[WKBackForwardNavigationDelegate webView:didFinishNavigation:]):
(-[WKBackForwardNavigationDelegate _webView:navigation:didSameDocumentNavigation:]):
(TEST):
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestNavigationDelegate.mm:
(-[TestNavigationDelegate _webView:navigation:didSameDocumentNavigation:]): Deleted.
(-[TestNavigationDelegate waitForDidFinishNavigationOrSameDocumentNavigation]): Deleted.
(-[WKWebView _test_waitForDidFinishNavigationOrSameDocumentNavigation]): Deleted.

Canonical link: https://commits.webkit.org/251894@main
